src/ops.rs

```diff
@@ -0,0 +1,411 @@
+                count.unwrap_or(50),
+                T::DTYPE.into(),
+                ctx,
+            ))
+        }
+    }
+
+    /// Repeat an array along a specified axis.
+    ///
+    /// Example:
+    ///
+    /// ```rust
+    /// use mlx::{array::Array, stream::StreamOrDevice};
+    /// // repeat a [2, 2] array 4 times along axis 1
+    /// let source = Array::from_slice(&[0, 1, 2, 3], &[2, 2]);
+    /// let r = Array::repeat::<i32>(source, 4, 1, StreamOrDevice::default());
+    /// ```
+    ///
+    /// # Parameters:
+    /// - array: array to repeat
+    /// - count: number of times to repeat
+    /// - axis: axis to repeat along
+    /// - stream: stream or device to evaluate on
+    pub fn repeat<T: ArrayElement>(
+        array: Array,
+        count: i32,
+        axis: i32,
+        stream: StreamOrDevice,
+    ) -> Array {
+        let ctx = stream.as_ptr();
+        unsafe { Array::from_ptr(mlx_sys::mlx_repeat(array.c_array, count, axis, ctx)) }
+    }
+
+    /// Repeat a flattened array along axis 0.
+    ///
+    /// Example:
+    ///
+    /// ```rust
+    /// use mlx::{array::Array, stream::StreamOrDevice};
+    /// // repeat a 4 element array 4 times along axis 0
+    /// let source = Array::from_slice(&[0, 1, 2, 3], &[2, 2]);
+    /// let r = Array::repeat_all::<i32>(source, 4, StreamOrDevice::default());
+    /// ```
+    ///
+    /// # Parameters:
+    /// - array: array to repeat
+    /// - count: number of times to repeat
+    /// - stream: stream or device to evaluate on
+    pub fn repeat_all<T: ArrayElement>(array: Array, count: i32, stream: StreamOrDevice) -> Array {
+        let ctx = stream.as_ptr();
+        unsafe { Array::from_ptr(mlx_sys::mlx_repeat_all(array.c_array, count, ctx)) }
+    }
+
+    /// An array with ones at and below the given diagonal and zeros elsewhere.
+    ///
+    /// Example:
+    ///
+    /// ```rust
+    /// use mlx::{array::Array, stream::StreamOrDevice};
+    /// // [5, 5] array with the lower triangle filled with 1s
+    /// let r = Array::tri::<f32>(5, None, None, StreamOrDevice::default());
+    /// ```
+    ///
+    /// # Parameters:
+    /// - n: number of rows in the output
+    /// - m: number of columns in the output -- equal to `n` if not specified
+    /// - k: index of the diagonal -- defaults to 0 if not specified
+    /// - stream: stream or device to evaluate on
+    pub fn tri<T: ArrayElement>(
+        n: i32,
+        m: Option<i32>,
+        k: Option<i32>,
+        stream: StreamOrDevice,
+    ) -> Array {
+        let ctx = stream.as_ptr();
+
+        unsafe {
+            Array::from_ptr(mlx_sys::mlx_tri(
+                n,
+                m.unwrap_or(n),
+                k.unwrap_or(0),
+                T::DTYPE.into(),
+                ctx,
+            ))
+        }
+    }
+
+    /// Element-wise absolute value.
+    ///
+    /// # Parameters:
+    /// - stream: stream or device to evaluate on
+    pub fn abs(&self, stream: StreamOrDevice) -> Array {
```

Comment:
Not specific to `abs()`, but for other functions, we should probably test whether it should take reference or ownership, since a lot of the ops impl moves the pointer with `std::move`.

Reply:
Did!